elastic/kibana
# Change updateObjectsSpaces API to prevent multiple objects w/ same origin

Resolves #127054.

### Background

When a shareable saved object is copied to another space, its ID changes but its "origin" is preserved in the `originId` field.

We need to prevent users from copying a saved object to one space, and then sharing the object to the same space. In other words, we want to prevent _two copies of an object that share an origin_ from existing in the same space. See the linked issue above for more details.

### Changes

The `SavedObjectClient.collectMultiNamespaceReferences` API takes a list of saved objects and collects the reference graph for each one, returning the reference graph as the result.

This PR changes that API so that for each reference, it checks all spaces for saved objects with matching origins, and it returns that result to the consumer.
The "Share to spaces" flyout now uses this information to prevent the user from sharing the object to those spaces.

The changes also include unit tests and API integration tests.
Finally, this PR adjusts the UI copy for consistency:
* "Assign spaces" -> "Share to spaces"
* "Copy to space" -> "Copy to spaces"

### Testing

<details><summary>Click to see script to create 26 spaces</summary>

```sh
USERNAME="elastic"
PASSWORD="changeme"
HOST="http://localhost:5601" # make sure to include your basepath, if any

declare -a spaceNames=("Alpha" "Bravo" "Charlie" "Delta" "Echo" "Foxtrot" "Golf" "Hotel" "India" "Juliet" "Kilo" "Lima" "Mike" "November" "Oscar" "Papa" "Quebec" "Romeo" "Sierra" "Tango" "Uniform" "Victor" "Whiskey" "X-Ray" "Yankee" "Zulu")
for name in "${spaceNames[@]}"
do
  id=`echo "$name" | awk '{print tolower($0)}'`
  curl -k -u $USERNAME:$PASSWORD -X POST "$HOST/api/spaces/space" -H 'kbn-xsrf: true' -H 'Content-Type: application/json' -d "{\"id\": \"$id\", \"name\": \"$name\"}"
done
```

</details>

1. Start Elasticsearch and Kibana
2. Add extra spaces to Kibana -- you can use a script to automate this process (see above)
3. From the Home page, add the eCommerce sample data
4. Navigate to Stack Management -> Saved Objects, find the sample data view, then open the "Share to spaces" flyout
   ![image](https://user-images.githubusercontent.com/5295965/164086604-589bf1c3-0c9d-4b2e-bdc4-bfda6ef5e7e3.png)
5. Observe that you can select any space and you can choose the "All spaces" option at the top:
   ![image](https://user-images.githubusercontent.com/5295965/164086858-7dbd61d4-ace9-4602-abc2-9546658ad846.png)
6. Share the data view to the Alpha space
7. Open the "Copy to spaces" flyout and copy the data view to the Bravo space, choosing the option to "Check for existing objects" (this preserves the origin)
8. Open the "Share to spaces" flyout again and observe that you can't share this object to "All spaces" or the Bravo space:
   ![image](https://user-images.githubusercontent.com/5295965/164087032-d7cd67b9-fe26-40b0-9bb2-6210569521cf.png)
   ![image](https://user-images.githubusercontent.com/5295965/164087123-5c245994-df43-40f3-bd27-4a2fcfe4ba3e.png)

### Other notes

This is not bullet-proof, there is a race condition where a user could view the "Share to spaces" flyout (step 5 above), _then_ someone else could copy the object to another space (step 7 above), and the flyout would not prevent the first user from sharing the object to that space. This is simply intended to minimize the chances that a user could get into the situation where multiple saved objects with the same origin exist in the same space.
